File: Sample_Questions/Practice_Questions/problem6/Python/problem6_attempt20.py
#Now sort the products data sorted by product price per category, use productcategoryid colunm to group by category
from pyspark import SparkConf, SparkContext
from pyspark.sql import SQLContext

conf = SparkConf().setAppName("problem6 Atttempt 20")
sc = SparkContext(conf=conf)
sqlContext =  SQLContext(sc)
# Approach 1 : Using DF

rdd = sc.textFile("hdfs://quickstart.cloudera:8020/user/cloudera/p93_products_sub_2/pa*")
from pyspark.sql import Row
product = Row('product_id','product_category_id','product_name','product_description','product_price','product_image')
dataPrepBeforeDF = rdd.map(lambda l : l.split(",")).map(lambda a : product(int(a[0]),int(a[1]),a[2],a[3],a[4],a[5]))
productsDF = dataPrepBeforeDF.toDF()

## Data Frame Create Apporach 2
from pyspark.sql.types import *
columns = [ StructField("product_id",IntegerType(),True),StructField("product_category_id",IntegerType(),True),StructField("product_name",StringType(),True),StructField("product_description",StringType(),True), StructField("product_price",FloatType(),True), StructField("product_image",StringType(),True)]
dataPrepBeforeDF2 = rdd.map(lambda l : l.split(",")).map(lambda a : Row(int(a[0]),int(a[1]),a[2],a[3],a[4],a[5]))
productsDF2 = sqlContext.createDataFrame(dataPrepBeforeDF2, columns)

## Using collect by list
productsDF.groupBy("product_category_id").agg(collect_list("product_id"))
#NameError: name 'collect_list' is not defined


# Approach 2 : Using RDD
## Use groupByKey 
rdd = sc.textFile("hdfs://quickstart.cloudera:8020/user/cloudera/p93_products/pa*")
## Create a Map RDD for key : product_category_id ; Value : Tuple/Array (product_id,product_price )
pairedRDD2 = rdd.map(lambda l : l.split('~') ).map(lambda a : (a[1], (a[0],a[4])))
## Create a group by key
groupedRDD2 = pairedRDD2.groupByKey()
## Sort iterables returned after groupByKey - specifyin the sort key
sortedRDD2 = groupedRDD2.mapValues(lambda v : sorted(list(v), key = lambda l : l[1]))
##Save results
sortedRDD2.saveAsTextFile("hdfs://quickstart.cloudera:8020/user/cloudera/spark_experiments/sortedRDDproblem6py")

##Sample data
## (u'4', [(u'72', u'149.99'), (u'59', u'159.99'), (u'62', u'179.97'), (u'69', u'179.97'), (u'66', u'1799.99'), (u'65', u'199.99'), (u'63', u'209.99'), (u'64', u'21.99'), (u'52', u'249.97'), (u'57', u'28.0'), (u'53', u'29.99'), (u'67', u'29.99'), (u'49', u'299.98'), (u'58', u'299.99'), (u'61', u'299.99'), (u'68', u'309.99'), (u'71', u'349.98'), (u'55', u'39.99'), (u'50', u'59.98'), (u'51', u'69.99'), (u'70', u'79.99'), (u'54', u'99.0'), (u'56', u'99.95'), (u'60', u'999.99')])
##(u'8', [(u'155', u'134.99'), (u'150', u'149.99'), (u'148', u'199.99'), (u'147', u'21.99'), (u'165', u'21.99'), (u'151', u'24.97'), (u'167', u'25.99'), (u'146', u'28.0'), (u'149', u'29.99'), (u'158', u'29.99'), (u'153', u'299.99'), (u'160', u'31.97'), (u'168', u'32.0'), (u'156', u'34.99'), (u'166', u'34.99'), (u'159', u'39.99'), (u'162', u'399.99'), (u'161', u'44.99'), (u'157', u'45.0'), (u'164', u'45.0'), (u'145', u'59.98'), (u'154', u'69.99'), (u'163', u'69.99'), (u'152', u'99.98')])
##(u'57', [(u'1284', u'0.0'), (u'1293', u'109.99'), (u'1294', u'134.99'), (u'1281', u'159.99'), (u'1289', u'159.99'), (u'1297', u'174.99'), (u'1279', u'189.99'), (u'1276', u'29.99'), (u'1295', u'29.99'), (u'1283', u'34.99'), (u'1280', u'39.99'), (u'1286', u'44.99'), (u'1292', u'59.97'), (u'1287', u'59.99'), (u'1288', u'59.99'), (u'1291', u'70.0'), (u'1275', u'75.0'), (u'1296', u'9.99'), (u'1274', u'90.0'), (u'1277', u'90.0'), (u'1278', u'90.0'), (u'1282', u'90.0'), (u'1285', u'90.0'), (u'1290', u'99.99')])

File: Sample_Questions/Practice_Questions/problem6/Scala/problem6-attempt20.scala
//Now sort the products data sorted by product price per category, use productcategoryid colunm to group by category
import org.apache.spark._
import org.apache.spark.sql._
import org.apache.spark.sql.types._

val conf = new SparkConf().setAppName("problem6 Atttempt 20")
val sc = new SparkContext(conf)
val sqlContext = new SQLContext(sc)


val rdd = sc.textFile("hdfs://quickstart.cloudera:8020/user/cloudera/p93_products_sub_2/pa*")

// Approach 1 : Using DF -  But to correct completely
// Data Frame Create Apporach 1
case class Person(product_id: Int, product_category_id: Int, product_name: String, product_description: String,product_price: Float,product_image: String)
val dataPrepBeforeDF = rdd.map(l => l.split(',')).map(a => Person(a(0).trim.toInt, a(1).trim.toInt,a(2),a(3), a(4).trim.toFloat,a(5)))
val productDF = dataPrepBeforeDF.toDF

// Data Frame Create Apporach 2
val structFields =  new Array[StructField](6)
structFields(0) = StructField("product_id",IntegerType,true)
structFields(1) = StructField("product_category_id",IntegerType,true)
structFields(2) = StructField("product_name",StringType,true)
structFields(3) = StructField("product_description",StringType,true)
structFields(4) = StructField("product_price",FloatType,true)
structFields(5) = StructField("product_image",StringType,true)
val schema =  new StructType(structFields)

// substep - Create RDD of Row
import org.apache.spark.sql.Row
val rowRDD = rdd.map(_.split(",")).map(a => Row(a(0).trim.toInt,a(1).trim.toInt,a(2),a(3),a(4).trim.toFloat,a(5)))
val productDF2 = sqlContext.createDataFrame(rowRDD,schema)


// Problem Statement Solution  Apprach 1
val solution1 = productDF.groupBy("product_category_id").sum("product_price").show

productDF.groupBy("product_category_id").agg(Map("product_price" -> "sum", "product_price" -> "avg")).orderBy(sum("product_price")).collect
orderBy("product_price")

import org.apache.spark.sql.functions._

// Using collect by list
productDF.groupBy("product_category_id").collect_List("product_price").orderBy(sum("product_price")).collect
//<console>:39: error: value collect_List is not a member of org.apache.spark.sql.GroupedData
//              productDF.groupBy("product_category_id").collect_List("product_price")


// Approach 2 : Using RDD
//Create paredRDD with key as - product_category_id 

// Use groupByKey 
val rdd = sc.textFile("hdfs://quickstart.cloudera:8020/user/cloudera/p93_products/pa*")
// Create a Map RDD for key : product_category_id ; Value : Tuple/Array (product_id,product_price )
val pairedRDD2 = rdd.map(l => (l.split("~")(1), (l.split("~")(0), l.split("~")(4))))
// Create a group by key
val groupedRDD2 = pairedRDD2.groupByKey()
// Sort iterables returned after groupBy Key
val sortedRDD2 = groupedRDD2.mapValues(v => v.toList.sortBy((_._2)))
//Save results
sortedRDD2.saveAsTextFile("hdfs://quickstart.cloudera:8020/user/cloudera/spark_experiments/sortedRDDproblem6")

/*
Sample
(4,List((72,149.99), (59,159.99), (62,179.97), (69,179.97), (66,1799.99), (65,199.99), (63,209.99), (64,21.99), (52,249.97), (57,28.0), (53,29.99), (67,29.99), (49,299.98), (58,299.99), (61,299.99), (68,309.99), (71,349.98), (55,39.99), (50,59.98), (51,69.99), (70,79.99), (54,99.0), (56,99.95), (60,999.99)))
(8,List((155,134.99), (150,149.99), (148,199.99), (147,21.99), (165,21.99), (151,24.97), (167,25.99), (146,28.0), (149,29.99), (158,29.99), (153,299.99), (160,31.97), (168,32.0), (156,34.99), (166,34.99), (159,39.99), (162,399.99), (161,44.99), (157,45.0), (164,45.0), (145,59.98), (154,69.99), (163,69.99), (152,99.98)))
(57,List((1284,0.0), (1293,109.99), (1294,134.99), (1281,159.99), (1289,159.99), (1297,174.99), (1279,189.99), (1276,29.99), (1295,29.99), (1283,34.99), (1280,39.99), (1286,44.99), (1292,59.97), (1287,59.99), (1288,59.99), (1291,70.0), (1275,75.0), (1296,9.99), (1274,90.0), (1277,90.0), (1278,90.0), (1282,90.0), (1285,90.0), (1290,99.99)))
*/

